content/en/blog/_posts/2021-XX-XX-writing-a-controller-for-pod-labels.md

```diff
@@ -0,0 +1,468 @@
+all you will need is `kubectl`. If you don't have a Kubernetes cluster at hand,
+I recommend you start one locally with [KinD (Kubernetes in Docker)][kind].
+
+All it takes to run the operator from your machine is this command:
+
+```bash
+make run
+```
+
+After a few seconds, you should see the operator's logs. Notice that our
+controller's `Reconcile` method was called for all Pods already running in the
+cluster.
+
+Let's keep the operator running and, in another terminal, create a new Pod:
+
+```bash
+kubectl run --image=nginx my-nginx
+```
+
+The operator should quickly print some logs, indicating that it reacted to the
+Pod's creation and subsequent changes in status:
+
+```text
+INFO    controllers.Pod no update required  {"pod": "default/my-nginx"}
+INFO    controllers.Pod no update required  {"pod": "default/my-nginx"}
+INFO    controllers.Pod no update required  {"pod": "default/my-nginx"}
+INFO    controllers.Pod no update required  {"pod": "default/my-nginx"}
+```
+
+Lets check the Pod's labels:
+
+```terminal
+$ kubectl get pod my-nginx --show-labels
+NAME       READY   STATUS    RESTARTS   AGE   LABELS
+my-nginx   1/1     Running   0          11m   run=my-nginx
+```
+
+Let's add an annotation to the Pod so that our controller knows to add our
+dynamic label to it:
+
+```bash
+kubectl annotate pod my-nginx padok.fr/add-pod-name-label=true
+```
+
+Notice that the controller immediately reacted and produced a new line in its
+logs:
+
+```text
+INFO    controllers.Pod adding label    {"pod": "default/my-nginx"}
+```
+
+```terminal
+$ kubectl get pod my-nginx --show-labels
+NAME       READY   STATUS    RESTARTS   AGE   LABELS
+my-nginx   1/1     Running   0          13m   padok.fr/pod-name=my-nginx,run=my-nginx
+```
+
+Bravo! You just successfully wrote a Kubernetes controller capable of adding
+labels with dynamic values to resources in your cluster.
+
+Controller and, more generally, operators both big and small, can be an
+important part of your Kubernetes journey. Writing operators is easier now than
+it has ever been. The possibilities are endless.
+
+## What next?
+
+If you want to go further, I recommend starting by deploying the operator inside
+a cluster. The `Makefile` generated by the Operator SDK will do most of the
+work.
+
+When deploying an operator to production, it is always a good idea to implement
+robust testing. The first step in that direction is to write unit tests.
+[This documentation][operator-sdk-testing] will guide you in writing tests for
+your operator. I wrote tests for the operator we just wrote; you can find all of
+my code in [this GitHub repository][github-repo].
+
+## How to learn more?
+
+The [Operator SDK documentation][operator-sdk-docs] goes into detail on how you
+can go further and implement more complex operators.
+
+When built-in Kubernetes types are not enough to model your use-case, you will
+need to use [Custom Resource Definitions (CRDs)][custom-resource-definitions].
```

Comment:
I'd reword this to fit the context that the article explained writing a controller, but of course the Operator SDK is also good for writing operators.

Reply:
Done 👍